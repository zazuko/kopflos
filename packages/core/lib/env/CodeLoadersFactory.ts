import type { AnyPointer } from 'clownface'
import LoaderRegistry from 'rdf-loaders-registry'
import cjsLoader from 'rdf-loader-code/ecmaScript.js'
import esmLoader from 'rdf-loader-code/ecmaScriptModule.js'
import { isGraphPointer } from 'is-graph-pointer'

export class CodeLoadersFactory {
  private registry!: LoaderRegistry

  init() {
    this.registry = new LoaderRegistry()
    esmLoader.register(this.registry)
    cjsLoader.register(this.registry)
  }

  load<T>(code: AnyPointer): Promise<T> | T | undefined {
    if (!isGraphPointer(code)) {
      return undefined
    }

    return this.registry.load<T>(code)
  }

  static exports = ['load']
}
