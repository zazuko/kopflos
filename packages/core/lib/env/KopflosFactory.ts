import type { KopflosConfig } from '../Kopflos.js'

export interface KopflosFactory {
  readonly kopflos: {
    readonly config: KopflosConfig
  }
}

export default (config: KopflosConfig) => class implements KopflosFactory {
  get kopflos() {
    return {
      config: Object.freeze(config),
    }
  }

  static exports = ['kopflos']
}
