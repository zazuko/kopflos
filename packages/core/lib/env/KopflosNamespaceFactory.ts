import type { NamespaceBuilder } from '@rdfjs/namespace'
import type { NamespaceFactory } from '@rdfjs/namespace/Factory.js'
import type { Environment } from '@rdfjs/environment/Environment.js'
import type NsBuildersFactory from '@tpluscode/rdf-ns-builders'
import { code } from '@zazuko/vocabulary-extras-builders'

type KopflosTerms = 'api' | 'Config' | 'Api' | 'resourceLoader' | 'handler' | 'method'

declare module '@tpluscode/rdf-ns-builders' {
  interface CustomNamespaces {
    kopflos: NamespaceBuilder<KopflosTerms>
    code: typeof code
  }
}

export class KopflosNamespaceFactory {
  init(this: Environment<NamespaceFactory | NsBuildersFactory>) {
    this.ns = {
      ...this.ns,
      kopflos: this.namespace('https://kopflos.described.at/'),
      code,
    }
  }
}
