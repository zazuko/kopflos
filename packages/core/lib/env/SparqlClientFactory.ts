import type { Options, StreamClient as IStreamClient } from 'sparql-http-client/StreamClient.js'
import StreamClient from 'sparql-http-client/StreamClient.js'
import type { ParsingClient as IParsingClient } from 'sparql-http-client/ParsingClient.js'
import ParsingClient from 'sparql-http-client/ParsingClient.js'

export interface Clients {
  stream: IStreamClient
  parsed: IParsingClient
}

export interface SparqlClientFactory {
  sparql: Record<string, Clients>
}

export default (endpoints: Record<string, Options | Clients>) => class implements SparqlClientFactory {
  sparql!: Record<string, Clients>

  init() {
    this.sparql = Object.fromEntries(Object.entries(endpoints).map(([key, endpointOrClients]) => {
      if ('stream' in endpointOrClients) {
        return [key, endpointOrClients]
      }

      return [key, {
        stream: new StreamClient(endpointOrClients),
        parsed: new ParsingClient(endpointOrClients),
      }]
    }))
  }
}
