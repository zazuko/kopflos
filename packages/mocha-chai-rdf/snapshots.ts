/* eslint-disable @typescript-eslint/no-explicit-any */
import chai, { expect } from 'chai'
import { jestSnapshotPlugin } from 'mocha-chai-jest-snapshot'
import rdf from '@zazuko/env-node'

declare global {
  /* eslint-disable @typescript-eslint/no-namespace */
  namespace Chai {
    interface Assertion {
      canonical: Assertion
    }
  }
}

chai.use(jestSnapshotPlugin())

chai.use((_chai, utils) => {
  const toMatchSnapshot = (chai.Assertion.prototype as any).toMatchSnapshot

  utils.addMethod(chai.Assertion.prototype, 'toMatchSnapshot', function (this: Chai.Assertion, ...args: any[]) {
    const obj = (this as any)._obj

    if (utils.flag(this, 'dataset-canonical')) {
      // Custom behavior for dataset
      expect(rdf.dataset.toCanonical(obj)).toMatchSnapshot(...args)
    } else {
      // Original behavior
      toMatchSnapshot.apply(this, args)
    }
  })

  utils.addProperty(chai.Assertion.prototype, 'canonical', function (this: Chai.Assertion) {
    utils.flag(this, 'dataset-canonical', true)
  })
})
