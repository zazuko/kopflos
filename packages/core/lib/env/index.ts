import parent, { Environment } from '@zazuko/env-node'
import { KopflosNamespaceFactory } from './KopflosNamespaceFactory.js'
import { CodeLoadersFactory } from './CodeLoadersFactory.js'

const env = new Environment([
  KopflosNamespaceFactory,
  CodeLoadersFactory,
], { parent })

export default env
export type KopflosEnvironment = typeof env
