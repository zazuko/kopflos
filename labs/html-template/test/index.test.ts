import sinon from 'sinon'
import { expect, use } from 'chai'
import type { HandlerArgs } from '@kopflos-cms/core'
import { createEnv } from '@kopflos-cms/core/env.js' // eslint-disable-line import/no-unresolved
import rdf from '@zazuko/env-node'
import type { AnyPointer } from 'clownface'
import snapshots from 'mocha-chai-rdf/snapshots.js'
import bindTemplate from '../index.js'

describe('@kopflos-labs/html-template', function () {
  use(snapshots)

  it('uses core representation a default template data', async function () {
    // given
    const templateFunc = sinon.stub()
    const context = {
      subject: rdf.clownface()
        .namedNode('')
        .addOut(rdf.ns.rdf.type, rdf.namedNode('http://example.com/Foo')),
      env: createEnv({
        baseIri: 'http://example.com/',
        sparql: {
          default: 'http://example.com/query',
        },
      }),
    } as HandlerArgs
    const previousResponse = {
      status: 200,
      body: '<html><body><template target-class="Foo"></template></body></html>',
      headers: {
        'Content-Type': 'text/html',
      },
    }

    // when
    const handler = bindTemplate(templateFunc)
    await handler(context, previousResponse)

    // then
    const templateData: AnyPointer = templateFunc.firstCall.args[1].pointer
    expect(templateData.dataset).canonical.toMatchSnapshot()
  })

  it('fetches data with given arguments', async function () {
    // given
    const templateFunc = sinon.stub()
    const fetchData = sinon.stub().returns(({ env }: HandlerArgs) => {
      return env.clownface()
        .blankNode()
        .addOut(rdf.ns.rdf.type, env.namedNode('http://example.com/Foo'))
        .dataset
    })
    const context = {
      subject: rdf.clownface().namedNode(''),
      env: createEnv({
        baseIri: 'http://example.com/',
        sparql: {
          default: 'http://example.com/query',
        },
      }),
    } as HandlerArgs
    const previousResponse = {
      status: 200,
      body: '<html><body><template target-class="Foo"></template></body></html>',
      headers: {
        'Content-Type': 'text/html',
      },
    }

    // when
    const handler = bindTemplate(templateFunc, fetchData, 'arg1', 'arg2')
    await handler(context, previousResponse)

    // then
    expect(fetchData).to.have.been.calledWith('arg1', 'arg2')
    const templateData: AnyPointer = templateFunc.firstCall.args[1].pointer
    expect(templateData.dataset).canonical.toMatchSnapshot()
  })

  it('fetches data as stream', async function () {
    // given
    const templateFunc = sinon.stub()
    const fetchData = sinon.stub().returns(({ env }: HandlerArgs) => {
      return env.clownface()
        .blankNode()
        .addOut(rdf.ns.rdf.type, env.namedNode('http://example.com/Foo'))
        .dataset.toStream()
    })
    const context = {
      subject: rdf.clownface().namedNode(''),
      env: createEnv({
        baseIri: 'http://example.com/',
        sparql: {
          default: 'http://example.com/query',
        },
      }),
    } as HandlerArgs
    const previousResponse = {
      status: 200,
      body: '<html><body><template target-class="Foo"></template></body></html>',
      headers: {
        'Content-Type': 'text/html',
      },
    }

    // when
    const handler = bindTemplate(templateFunc, fetchData, 'arg1', 'arg2')
    await handler(context, previousResponse)

    // then
    const templateData: AnyPointer = templateFunc.firstCall.args[1].pointer
    expect(templateData.dataset).canonical.toMatchSnapshot()
  })

  it('returns error if previous response is not a string', async function () {
    // given
    const context = {
      env: createEnv({
        baseIri: 'http://example.com/',
        sparql: {
          default: 'http://example.com/query',
        },
      }),
    } as HandlerArgs
    const previousResponse = {
      status: 200,
      body: rdf.dataset(),
    }

    // when
    const handler = bindTemplate(sinon.stub())
    const result = await handler(context, previousResponse)

    // then
    expect(result).to.be.an('error')
  })

  it('returns error if there is no previous response', async function () {
    // given
    const context = {
      env: createEnv({
        baseIri: 'http://example.com/',
        sparql: {
          default: 'http://example.com/query',
        },
      }),
    } as HandlerArgs

    // when
    const handler = bindTemplate(sinon.stub())
    const result = await handler(context)

    // then
    expect(result).to.be.an('error')
  })
})
