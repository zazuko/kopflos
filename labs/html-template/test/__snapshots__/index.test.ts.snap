// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@kopflos-labs/html-template fetches data as stream 1`] = `
"_:c14n0 <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://example.com/Foo> .
"
`;

exports[`@kopflos-labs/html-template fetches data with given arguments 1`] = `
"_:c14n0 <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://example.com/Foo> .
"
`;
